client/src/stellar_error.rs

```diff
@@ -0,0 +1,244 @@
+use std::error::Error;
+use serde::{de, Deserialize, Deserializer};
+use std::fmt;
+
+/// A resource for the stellar horizon API specific error codes.
+/// These errors adhere to the Problem Details Standard
+/// https://tools.ietf.org/html/draft-ietf-appsawg-http-problem-00
+/// and a list of possible erros can be found here
+/// https://www.stellar.org/developers/horizon/reference/errors.html
+#[derive(Debug)]
+pub struct StellarError {
+    // type is a protected word
+    error_type: ErrorType,
+    title: String,
+    status: u16,
+    detail: String,
+    instance: String,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum ErrorType {
+    BadRequest,
+    BeforeHistory,
+    Forbidden,
+    NotAcceptable,
+    NotFound,
+    NotImplemented,
+    RateLimitExceeded,
+    InternalServerError,
+    StaleHistory,
+    TransactionFailed,
+    TransactionMalformed,
+    OtherError(String),
+}
+
+#[derive(Deserialize)]
+struct IntermediateStellarError {
+    // type is a protected word
+    #[serde(rename = "type")] error_type: String,
+    title: String,
+    status: u16,
+    detail: String,
+    instance: String,
+}
+
+impl<'de> Deserialize<'de> for StellarError {
+    fn deserialize<D>(d: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let rep: IntermediateStellarError = IntermediateStellarError::deserialize(d)?;
+        StellarError::new(
+            &rep.error_type,
+            rep.title,
+            rep.status,
+            rep.detail,
+            rep.instance,
+        ).map_err(|err| de::Error::custom(err))
+    }
+}
+
+impl StellarError {
+    /// Creates a new stellar error.  If the error type is not recognized an error is returned.
+    pub fn new(
```

Comment:
What's the advantage of this being public?

Reply:
bad answer but I use it in a test :/ I don't think an external lib would ever have a reason to call this though.